Change folder breadcrumb on folder change in general settings

mdvictor opened the pull request:
Fixes a bug where the folder breadcrumb no longer appears after changing the folder from a dashboards general settings. This happens because not all folders in the general settings select box are actually stored in the `navIndex`, so when this changes the dashboard scene `getPageNav` method doesn't find the new folder and returns `not-found`. This solution updates the `navIndex` to include selected folder in it so that the dashboard can pick it up easily and render it in the breadcrumbs.

Permissions are accounted for by the folders select box in the settings page so we don't need to worry about that.


https://github.com/grafana/grafana/assets/36818606/5a3ab752-c0dd-40ad-8b3b-0b87912c090b



Fixes https://github.com/grafana/grafana/issues/85547

mdvictor committed:
Change folder breadcrumb on folder change in general settings

ivanortegaalba commented on public/app/features/dashboard-scene/settings/GeneralSettingsEditView.tsx:
```diff
@@ -96,13 +97,18 @@ export class GeneralSettingsEditView
   };
 
-  public onFolderChange = (newUID: string | undefined, newTitle: string | undefined) => {
+  public onFolderChange = async (newUID: string | undefined, newTitle: string | undefined) => {
```
Great finding! I'd suggest adding some tests here to make sure we call the updateNavModel with the new folder ID. WDYT?

mdvictor replied:
adapted tests for this

ivanortegaalba commented on public/app/features/dashboard-scene/pages/DashboardScenePageStateManager.ts:
```diff
@@ -228,6 +228,15 @@ export class DashboardScenePageStateManager extends StateManagerBase<DashboardSc
     return null;
   }
 
+  public async updateNavModel(folderUid: string) {
```
What do you think about updating signature for [`initNavModel`](https://github.com/grafana/grafana/pull/86342/files#diff-014099493438b3f54a14bf4d12e496c65d45fc3dd9b8a91f27a0d1571f538015L231) to keep consistent the function signatures?

Also, I'd add some tests here. WDYT?

mdvictor replied:
In that case we might as well cut one of the functions and use the same everywhere, which is what I'm going for now :D

ivanortegaalba reviewed (approved):
It works like a charm! Good work Victor!

mdvictor committed:
tests and refactor

mdvictor committed:
Merge branch 'main' into mdvictor/fix-folder-breadcrumb-general-settings

mdvictor committed:
refactor to fix broken tests

mdvictor commented:
@ivanortegaalba I refactored things a bit. Calling the `DashboardScenePageStateManager` in the settings somehow caused all tests in `SaveDashboardDrawer.test.tsx` to fail because of some redux reducers being called badly at some point. Couldn't figure out why or how things connected tho so I moved the `updateNavModel` function in a `utils` file. If you have a better idea on keeping this in the state manager let me know!

mdvictor committed:
fix test
